### PR TITLE
layout: Remove cached thread local context from LayoutContext, and use LayoutContext for assign_inline_sizes()

### PR DESCRIPTION
**Pull request by stshine**
According to https://github.com/servo/servo/pull/3069 the cached thread local context is introduced for green threads. Now green threads has gone, and the existence of cache force us to create a `LayoutContext`, an `AssignISizes` and an `AssignBSizes` for each flow during parallel layout, so the pull request tries to remove it. And it also switch `assign_inline_sizes()` to accept a `LayoutContext` parameter, as according to my current design we need to do full layout to some flex items for column flexbox during assign isize traversal.

Part of #14123.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because refactoring

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15417)

### CODE REVIEW

**Comment by stshine** [comment]
The main concern of this pull request is that it changed how we access thread local layout context. Now we may need to access tls several times for each text fragment, compare to one time for each flow previously. But @pcwalton said tls access is very fast(three instructions), so this may be small. And it simplifies our parallel layout code, so it may have its value.

Would love to know how to make this better. Please criticise!

r? @pcwalton or @emilio

File: components/layout/context.rs [hunk 1]
```diff
@@ -50,53 +49,32 @@ impl<E: TElement> BorrowMut<ThreadLocalStyleContext<E>> for ScopedThreadLocalLay
...
-    LOCAL_CONTEXT_KEY.with(|r| {
-        let mut r = r.borrow_mut();
-        if let Some(context) = r.clone() {
-            context
-        } else {
-            let context = PersistentThreadLocalLayoutContext::new(shared);
-            *r = Some(context.clone());
-            context
+pub fn with_font_context<F, R>(layout_context: &LayoutContext, f: F) -> R
+    where F: FnOnce(&mut FontContext) -> R {
```
Comment (emilio) [inline components/layout/context.rs]: nit: Brace on a newline after where.

File: components/layout/context.rs [hunk 2]
```diff
@@ -50,53 +49,32 @@ impl<E: TElement> BorrowMut<ThreadLocalStyleContext<E>> for ScopedThreadLocalLay
...
-                                    -> Rc<PersistentThreadLocalLayoutContext> {
-    LOCAL_CONTEXT_KEY.with(|r| {
-        let mut r = r.borrow_mut();
-        if let Some(context) = r.clone() {
-            context
-        } else {
-            let context = PersistentThreadLocalLayoutContext::new(shared);
-            *r = Some(context.clone());
-            context
+pub fn with_font_context<F, R>(layout_context: &LayoutContext, f: F) -> R
```
Comment (emilio) [inline components/layout/context.rs]: Maybe rename to `with_thread_local_font_context` or something, so it's clear what's going on?

**Review by emilio — APPROVED** [review]
Looks great, thanks for the cleanup @stshine!

I don't feel too strongly about my two following comments, so feel free to land this addressing them or not.

**Comment by stshine** [comment]
Comments addressed. Thank you @emilio :)
But I give my main thanks to rustc, without you I would not dare to touch code like this.

@bors-servo r=emilio

**Pull request merged**